ENH: Improve the Tuple Count validation error reporting

imikejackson opened the pull request.

imikejackson committed:
ENH: Improve the Tuple Count validation error reporting

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Rewrite using nonstd::expected instead of std::pair<bool, string>

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

JDuffeyBQ commented on src/complex/DataStructure/DataStructure.hpp:
```diff
@@ -650,7 +651,7 @@ class COMPLEX_EXPORT DataStructure
    * @return bool
    */
-  bool validateNumberOfTuples(const std::vector<DataPath>& dataPaths) const;
+  nonstd::expected<bool, std::string> validateNumberOfTuples(const std::vector<DataPath>& dataPaths) const;
```
```suggestion
  nonstd::expected<void, std::string> validateNumberOfTuples(const std::vector<DataPath>& dataPaths) const;
```

JDuffeyBQ commented on src/complex/DataStructure/DataStructure.cpp:
```diff
@@ -775,13 +776,15 @@ void DataStructure::applyAllDataStructure()
 }
 
-bool DataStructure::validateNumberOfTuples(const std::vector<DataPath>& dataPaths) const
+nonstd::expected<bool, std::string> DataStructure::validateNumberOfTuples(const std::vector<DataPath>& dataPaths) const
```
```suggestion
nonstd::expected<void, std::string> DataStructure::validateNumberOfTuples(const std::vector<DataPath>& dataPaths) const
```

JDuffeyBQ commented on src/complex/DataStructure/DataStructure.cpp:
```diff
@@ -775,13 +776,15 @@ void DataStructure::applyAllDataStructure()
 {
   if(dataPaths.empty())
   {
     return true;
```
```suggestion
    return {};
```

JDuffeyBQ commented on src/complex/DataStructure/DataStructure.cpp:
```diff
@@ -806,7 +816,7 @@ bool DataStructure::validateNumberOfTuples(const std::vector<DataPath>& dataPath
+      return {nonstd::make_unexpected(message.str())};
     }
   }
   return true;
```
```suggestion
  return {};
```

JDuffeyBQ committed:
Apply suggestions from code review

Changed `nonstd::expected<bool, std::string>` to `nonstd::expected<void, std::string>`
